apiserver/facades/controller/agenttools/agenttools.go

```diff
@@ -80,10 +86,18 @@ func checkToolsAvailability(ctx context.Context, newEnviron newEnvironFunc, mode
 	// only return patches for the passed major.minor (from major.minor.patch).
 	// We'll try the released stream first, then fall back to the current configured stream
 	// if no released tools are found.
-	vers, err := finder(ctx, ss, env, currentVersion.Major, currentVersion.Minor, []string{tools.ReleasedStream}, coretools.Filter{})
-	preferredStream := tools.PreferredStreams(&currentVersion, modelCfg.Development(), modelCfg.AgentStream())[0]
-	if preferredStream != tools.ReleasedStream && errors.Cause(err) == coretools.ErrNoMatches {
-		vers, err = finder(ctx, ss, env, currentVersion.Major, currentVersion.Minor, []string{preferredStream}, coretools.Filter{})
+	vers, err := api.findTools(ctx, ss, env, currentVersion.Major, currentVersion.Minor, []string{tools.ReleasedStream}, coretools.Filter{})
+	if errors.Is(err, coretools.ErrNoMatches) {
+		var modelCfg *config.Config
+		modelCfg, err = api.modelConfigService.ModelConfig(ctx)
+		if err != nil {
+			return version.Zero, errors.Annotate(err, "cannot get config")
+		}
+
+		preferredStream := tools.PreferredStreams(&currentVersion, modelCfg.Development(), modelCfg.AgentStream())[0]
+		if preferredStream != tools.ReleasedStream {
+			vers, err = api.findTools(ctx, ss, env, currentVersion.Major, currentVersion.Minor, []string{preferredStream}, coretools.Filter{})
+		}
```

Comment:
Honestly I don't think this logic is correct, I think we should be getting the list of `preferredStreams` first from `tools.PreferredStreams`, and then feeding the whole list to `api.findTools`.

@wallyworld thoughts? Maybe I'm missing something here as I don't know why we did it this way in the first place.

Reply:
I think the original logic is clumsy and inefficient but the intent seems in line with what we want, ie search for agents in order of devel, proposed, released; first match wins. FindTools handles this - I suspect the code here was done earlier and was never refactored. As you say, we just pass into FindTools the list of streams to search, preferred ones first in the list; one call to FindTools.

Reply:
Great, I'll make this change. Technically it's a behaviour change, but I think it's bringing the behaviour more in line with what a user would expect, plus I doubt anyone's depending on the old behaviour